lib/cc-cubic.c

```diff
@@ -0,0 +1,156 @@
+ * IN THE SOFTWARE.
+ */
+
+#include <math.h>
+#include "quicly/cc.h"
+
+#define QUICLY_MIN_CWND 2
+
+typedef double cubic_float_t;
+#define QUICLY_CUBIC_C ((cubic_float_t)0.4)
+#define QUICLY_CUBIC_BETA ((cubic_float_t)0.7)
+
+/* Calculates the time elapsed since the last congestion event (parameter t) */
+static cubic_float_t calc_cubic_t(const quicly_cc_t *cc, int64_t now)
+{
+    cubic_float_t clock_delta = now - cc->state.cubic.avoidance_start;
+    return clock_delta / 1000; /* ms -> s */
+}
+
+/* Equation 1 (using bytes as unit instead of MSS) */
+static uint32_t calc_w_cubic(const quicly_cc_t *cc, cubic_float_t t_sec, uint32_t max_udp_payload_size)
+{
+    cubic_float_t tk = t_sec - cc->state.cubic.k;
+    return (QUICLY_CUBIC_C * (tk * tk * tk) * max_udp_payload_size) + cc->state.cubic.w_max;
+}
+
+/* Equation 2 */
+/* K depends solely on W_max, so we update both together on congestion events */
+static void update_cubic_k(quicly_cc_t *cc, uint32_t max_udp_payload_size)
+{
+    cubic_float_t w_max_mss = cc->state.cubic.w_max / (cubic_float_t)max_udp_payload_size;
+    cc->state.cubic.k = cbrt(w_max_mss * ((1 - QUICLY_CUBIC_BETA) / QUICLY_CUBIC_C));
+}
+
+/* Equation 4 (using bytes as unit instead of MSS) */
+static uint32_t calc_w_est(const quicly_cc_t *cc, cubic_float_t t_sec, cubic_float_t rtt_sec, uint32_t max_udp_payload_size)
+{
+    return (cc->state.cubic.w_max * QUICLY_CUBIC_BETA) +
+           ((3 * (1 - QUICLY_CUBIC_BETA) / (1 + QUICLY_CUBIC_BETA)) * (t_sec / rtt_sec) * max_udp_payload_size);
+}
+
+/* TODO: Avoid increase if sender was application limited. */
+static void cubic_on_acked(quicly_cc_t *cc, const quicly_loss_t *loss, uint32_t bytes, uint64_t largest_acked, uint32_t inflight,
+                           int64_t now, uint32_t max_udp_payload_size)
+{
+    assert(inflight >= bytes);
+    /* Do not increase congestion window while in recovery. */
+    if (largest_acked < cc->recovery_end)
+        return;
+
+    /* Slow start. */
+    if (cc->cwnd < cc->ssthresh) {
+        cc->cwnd += bytes;
+        if (cc->cwnd_maximum < cc->cwnd)
+            cc->cwnd_maximum = cc->cwnd;
+        return;
+    }
+
+    /* Congestion avoidance. */
+    cubic_float_t t_sec = calc_cubic_t(cc, now);
+    cubic_float_t rtt_sec = loss->rtt.smoothed / (cubic_float_t)1000; /* ms -> s */
+
+    uint32_t w_cubic = calc_w_cubic(cc, t_sec, max_udp_payload_size);
+    uint32_t w_est = calc_w_est(cc, t_sec, rtt_sec, max_udp_payload_size);
+
+    if (w_cubic < w_est) {
+        /* 4.2 TCP-Friendly Region */
+        /* Prevent cwnd from shrinking if W_est is reduced due to RTT increase */
+        if (w_est > cc->cwnd) {
+            cc->cwnd = w_est;
+            if (cc->cwnd_maximum < cc->cwnd)
+                cc->cwnd_maximum = cc->cwnd;
+        }
+    } else {
+        /* 4.3/4.4 CUBIC Region */
+        /* Prevent cwnd from shrinking after fast convergence (beta*W_max < cwnd) */
```

Comment:
How is this prevented in the code here?

Reply:
Through the check for `increment > cc->cwnd`. Otherwise, for t close to 0 after fast convergence, `W_cubic(t) = beta * W_max = beta * 0.85 * W_last_max` < `beta * W_last_max = cwnd`.

Similarly, W_est(t) can shrink when the estimated RTT changes, hence there is a similar check there.

Reply:
Extended the comment now.